src/transformStateWithClones.js

```diff
@@ -7,7 +7,30 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        copyState = { ...copyState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete copyState[key];
+        }
+        break;
+
+      case 'clear':
+        copyState = {};
+        break;
+    }
+
+    stateHistory.push({ ...copyState });
```

Comment:
The `stateHistory.push({ ...copyState });` operation should be moved outside of the `switch` block, but still within the loop. This ensures that the state is recorded after processing each action, as per the additional prompt rules.